File: Python/search/KokoBananas.py
"""
Demonstrates the case when you need to find the minimum number for something rather than the number itself.
One approach is to keep decreasing when even the value below it is lower.
This one offers a faster solution. Keep setting the res and if that value happens to be the lowest, the l will increment up.
"""

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles)
        res = r
    

        while (l <= r):
            k, hours = l + ((r - l)//2), 0

            for p in piles:
                hours += math.ceil(float(p) / k)

            if hours > h:
                l = k + 1
            else:
                res = k
                r = k - 1
                
        return res

File: c++/questions/KokoBananas.cpp
#include <vector>
using namespace std;

// note that h has to be greater than the number of piles. Hence, the max k can be is the max element of the piles.

class Solution
{
public:
    int minEatingSpeed(vector<int> &piles, int h)
    {
        int l = 1, r = *max_element(piles.begin(), piles.end()), res = 0; // because hours is greater than or equal to piles

        while (l <= r)
        {
            int hours = 0;
            int k = l + ((r - l) / 2);

            for (auto pile : piles)
            {
                hours += ceil((double)pile / k);
            }

            if (hours > h)
            {
                l = k + 1;
            }
            else
            {
                res = k;
                r = k - 1;
            }
        }
        return res;
    }
};
